ruiyangzh iP

ruiyangzh opened the pull request:
# Luke
## It's Duke but **better** or something ©

>"Rarely is the question asked: Is our children learning?" - George W. Bush

release [here.](https://github.com/ruiyangzh/ip/releases/tag/A-Jar)
- Saves to disk
- extensible support for more tasks and commands
- "help" command for <sub>noobs</sub> 🤔 


- [x] Generic command interface
- [x] Task parent class
- [ ] JavaFX UI (TBD)

ok i ran out of ideas now here's a number list of *top numbers*

1. 3
1. 1
1. 2

```java
public class Main {
    public static void main(String[] args) {
       System.out.println("Goodbye world!")
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ruiyangzh committed:
Stage 1; renamed bot to Luke. Gave it basic startup and end lines.

ruiyangzh committed:
Stage 1; added echo for user input, implemented using scanner.

ruiyangzh committed:
Level 2. Added list functionality (implemented using an array of String).

ruiyangzh committed:
Level 3. Implemented Task class, added ability to mark tasks as done or undone. Restructured the holding array from String[] to Task[].

ruiyangzh committed:
Added Deadline, Event and ToDo classes that extend Task.

ruiyangzh committed:
Restructured Duke; with a processInput that handles scanner inputs.
Implemented adding each task variant into the list of Duke.

ruiyangzh committed:
Implemented A-TextUiTesting, with defined input/output.

ruiyangzh committed:
Implemented A-Exceptions, partway through Level 5.

ruiyangzh committed:
Implemented Level 5, exceptions added for invalid user inputs.

ruiyangzh committed:
Implemented A-Collections, using an arraylist to store tasks.

ruiyangzh committed:
Implemented Level-6, added delete functionality.

ruiyangzh committed:
Implemented Level 7.

ruiyangzh committed:
Implemented Level 8. TODO-Handle parsing errors.

ruiyangzh committed:
Partially implement OOP and packages - not complete.

ruiyangzh committed:
Partially implement OOP and packages - not complete.

ruiyangzh committed:
Implement A-MoreOOP and A-Packages.

ruiyangzh committed:
Merge branch 'branch-OOP-and-Packages' into add-gradle-support

ruiyangzh committed:
Implement A-Gradle. Partial stylecheck compliance.

ruiyangzh committed:
Implement A-JUnit. Package as JAR file, in releases.(Implement A-Jar)

ruiyangzh committed:
Implement Level-9.

ruiyangzh committed:
Implement A-JavaDoc.

ruiyangzh committed:
Implement A-CodingStandard.

ruiyangzh committed:
Merge branch 'branch-A-CodingStandard'

ruiyangzh committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/TaskList.java

ruiyangzh committed:
Miscellaneous merge fixing.

ruiyangzh committed:
In-progress implementation of JavaFX.

tanboonkhong commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,105 @@
+     * clearing the file or shutting down.
+     * @return whether the handler decides to call for a shutdown of the Duke instance.
+     */
+    public boolean corruptedFileHandler() {
```
The method returns a boolean, consider renaming the method to make it sounds like a boolean.

tanboonkhong commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,105 @@
+        }
+        luke.closeDuke();
+    }
+}
```
The main method seems a bit lengthy. Perhaps can abstract out parts into smaller parts to improve readability.

tanboonkhong commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,171 @@
+                break;
+            case ("event"):
+                int split1 = input.indexOf("/from");
+                int split2 = input.indexOf("/to");
```
The naming of split1 and split2 might be confusing, perhaps giving it a meaningful name for make it clearer?

tanboonkhong commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,165 @@
+        if (targetIndex > counter || targetIndex < 0) {
+            return false;
+        }
+        if (isToBeMarked) {
+            taskList.get(targetIndex).markDone();
+        }
+        if (!isToBeMarked) {
+            taskList.get(targetIndex).markUndone();
+        }
```
Would it be better to use 'if-else' rather than both if cases?

tanboonkhong reviewed (commented):
Overall, the codebase is structured well, and adheres to OOP principles. There are some minor issues regarding the naming of variables and coding styles.

KamiliArsyad commented on src/main/java/command/Commandable.java:
```diff
@@ -0,0 +1,15 @@
+ * if the execution causes the program to shut down. Also able to throw a FailureInExecuteException if the execution
+ * happens to fail for one reason or another.
+ */
+public interface Commandable {
```
While it is a common belief that interfaces should have `-able` suffix in their names, I think it is not required to adhere this especially if it does not provide a semantic meaning. Maybe naming it into "Executable" or "CommandExecutable" makes more sense if you're sticking with this rule?

KamiliArsyad commented on src/main/java/command/HelpCommand.java:
```diff
@@ -0,0 +1,31 @@
+                + "delete: delete(int x); deletes the indicated task\n"
+                + "find: find(String keyword); finds any tasks that contain the keyword\n"
+        );
+        return false;
```
Alternatively, add a help-related method for Commandable so that each implementation may add their own help independently. Then this help command may use the TaskList?

KamiliArsyad reviewed (commented):
Nicely made and organized! It's apparent that there's a lot of thought process going in this project.

dhruvir29 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,171 @@
+     * @throws InvalidCommandException if the command cannot be identified.
+     * @throws InvalidVarException if the command is identifiable but the parameters are incorrect.
+     */
+    public Commandable parse(String input) throws InvalidCommandException, InvalidVarException {
```
Generally, it seems that it is not recommended for methods to be too long (more than 30 lines). Perhaps some of the code could be abstracted out!

dhruvir29 reviewed (commented):
Overall your code quality is quite good as you have segmented it well

ruiyangzh committed:
Implement Level-10.

ruiyangzh committed:
Commandable: Refactor naming to Executable

Commandable is not a very suitable name for this interface, since it is not a proper english word and does not represent what it does. Since the shared method is named execute(), this schema makes more sense.

Replace all instances of the word "Commandable" with the word "Executable", for a more accurate representation of the name.

ruiyangzh committed:
In-progress for A-Assertions

ruiyangzh committed:
Add JavaDocs for public methods

Previously, most methods were missing JavaDocs documentations. This may hinder development for those who are new to the project.

Javadocs are to be added to most public methods.

ruiyangzh committed:
Add assertions in various places

Assertions were not present in earlier code, but adding them in could improve code checking by spotting errors early.

Add assertions into the bot (some are done in previous commits.)

ruiyangzh committed:
Inprogress refactoring of Parser

ruiyangzh committed:
Refactor Parser and UserInterface

Parser and UserInterface had very long methods in prior commits, for example sceneMaker() and parse(). This made the code difficult to understand.

In adherence with code quality standards, refactor and split up the two mentioned methods to increase readability and clarity.

ruiyangzh committed:
In progress refactoring of TaskList

ruiyangzh committed:
TaskList : refactor stringListToTaskList

TaskList had a very unwieldy method stringListToTaskList that was difficult to understand.

Refactor the method and split it up into smaller methods.

ruiyangzh committed:
Merge pull request #1 from ruiyangzh/branch-A-Assertions

Add Assertions into various points in code

ruiyangzh committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/command/AddTaskExecutable.java
#	src/main/java/command/DeleteExecutable.java
#	src/main/java/command/FindCommand.java
#	src/main/java/command/MarkExecutable.java
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/UserInterface.java

ruiyangzh committed:
Merge remote-tracking branch 'origin/branch-A-CodeQuality' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/task/Task.java

ruiyangzh committed:
Merge conflict resolution

ruiyangzh committed:
Merge pull request #2 from ruiyangzh/branch-A-CodeQuality

Code quality refactoring

ruiyangzh committed:
Add 2 new commands, rebind and unbind

Added the ability to rebind commands to a custom string (must not contain whitespaces), and use them as shorthand. The binds can also be removed.

ruiyangzh committed:
Refactor Javadoc comments

ruiyangzh committed:
Refactor Javadoc comments

ruiyangzh committed:
Bugfix binding commands

ruiyangzh committed:
Bugfix TaskList and UserInterface

ruiyangzh committed:
Prepare README.md and documentation

ruiyangzh committed:
Fix tests failing due to recent changes

ruiyangzh committed:
Fix tests failing due to recent changes

ruiyangzh committed:
Bugfix and add Ui.png

ruiyangzh committed:
Minor fix to README.md

ruiyangzh committed:
Bugfix and credits

ruiyangzh committed:
Formatting

ruiyangzh committed:
Added check for chronological events

ruiyangzh committed:
Added documentation for the check
